lyse/__init__.py

```diff
@@ -296,6 +299,102 @@ def h5_path(self):
     def no_write(self):
         """bool: The value provided for `no_write` during instantiation."""
         return self.__no_write
+    
+    @property
+    def h5_file(self):
+        """h5py.File: opened h5py file handle for the shot"""
+        return self.__h5_file
+    
+    @contextlib.contextmanager
+    def open(self, mode):
+        """Opens the Run's h5 file for successive reads/writes.
+
+        Intended to be used as a context manager.
+        Supports all h5 modes, though only `'r'` and `'r+'`/`'a'` are typically
+        used within lyse itself.
+
+        Args:
+            mode (str): which mode to open the h5 file with.
+                Must be 'r', 'a', 'r+', 'w', 'w-', or 'x'.
+                Lyse typically only uses 'r' and 'a'.
+
+        Examples:
+            >>> from lyse import *
+            >>> shot = Run(path)
+            >>> with shot.open('a'):
+            >>>     # shot processing that requires reads/writes
+            >>>     t, vals = shot.get_trace('my_trace')
+            >>>     results = vals**2
+            >>>     shot.save_result_array('my_result', results)
+            >>> # Shot processing that doesn't require h5 reads/writes
+            >>> print(f'Max of results is {results.max():.3f}')
+            Max of results is 5.003
+
+        Raises:
+            PermissionError: If the Run is set as read-only but a write mode is requested
+        """
+
+        # ensure no_write is respected
+        if self.no_write:
+            if mode == 'r':
+                pass
+            else:
+                msg = f'Cannot open file with a write mode; this run is read-only'
+                raise PermissionError(msg)
```

Comment:
This could be condensed to `if self.no_write and mode != “r”:`

Reply:
Well noted! This is the result of simplifying the even more convoluted logic tree I started with.